File: msgpack_play/proto_gen/t_proto.cs
//Dependent library  https://github.com/neuecc/MessagePack-CSharp
using System;
using System.Collections;
using System.Collections.Generic;
using MessagePack;

//账号信息
[MessagePackObject]
public class UserInfo {
    //账号ID
    [Key(0)] public int uid;
    //账号名称
    [Key(1)] public string nick;
    //标签
    [Key(2)] public List<string> tags;
    //性别
    [Key(3)] public byte sex;
    //年纪
    [Key(4)] public byte age;
    //声音设置
    [Key(5)] public boolean sound;
}

//登录请求
[MessagePackObject]
public class LoginReq {
    //账号
    [Key(0)] public string userName;
    //校验码
    [Key(1)] public string code;
}

//登录响应
[MessagePackObject]
public class LoginRsp {
    //状态码
    [Key(0)] public short code;
    //状态信息
    [Key(1)] public string msg;
    //用户信息
    [Key(2)] public UserInfo data;
}

//房间
[MessagePackObject]
public class Room {
    [Key(0)] public long rid;
    [Key(1)] public string desc;
    [Key(2)] public boolean playing;
    [Key(3)] public DateTime create;
    [Key(4)] public HashSet<UserInfo> players;
    [Key(5)] public Dictionary<int, UserInfo> map1;
    [Key(6)] public Dictionary<int, UserInfo> map2;
    [Key(7)] public byte[] log;
}
public class _ID_Map
{
    public static Dictionary<System.Type, int> Type2Id = new Dictionary<System.Type, int>();
    public static Dictionary<int, System.Type> Id2Type = new Dictionary<int, System.Type>();
    public static void Init()
    {
        _ID_Map.Type2Id[typeof(UserInfo)] = 1;
        _ID_Map.Id2Type[1] = typeof(UserInfo);
        _ID_Map.Type2Id[typeof(LoginReq)] = 2;
        _ID_Map.Id2Type[2] = typeof(LoginReq);
        _ID_Map.Type2Id[typeof(LoginRsp)] = 3;
        _ID_Map.Id2Type[3] = typeof(LoginRsp);
        _ID_Map.Type2Id[typeof(Room)] = 4;
        _ID_Map.Id2Type[4] = typeof(Room);
    }
}

File: msgpack_play/proto_gen/t_proto.ts
import {  MsgArray, MType } from "msgpack_play";

/**账号信息*/
@MsgArray.Meta({
    id: 1,
    name: "UserInfo",
    fields: [
        ["uid", MType.I32, 1],
        ["nick", MType.STR, 1],
        ["tags", ["Arr", MType.STR], 0],
        ["sex", MType.I8, 0],
        ["age", MType.I8, 0],
        ["sound", MType.BOOL, 0]
    ]
})
export class UserInfo {
    /**账号ID*/
    public uid: number;
    /**账号名称*/
    public nick: string;
    /**标签*/
    public tags: Array<string>;
    /**性别*/
    public sex: number;
    /**年纪*/
    public age: number;
    /**声音设置*/
    public sound: boolean;
}

/**登录请求*/
@MsgArray.Meta({
    id: 2,
    name: "LoginReq",
    fields: [
        ["userName", MType.STR, 1],
        ["code", MType.STR, 1]
    ]
})
export class LoginReq {
    /**账号*/
    public userName: string;
    /**校验码*/
    public code: string;
}

/**登录响应*/
@MsgArray.Meta({
    id: 3,
    name: "LoginRsp",
    fields: [
        ["code", MType.I16, 1],
        ["msg", MType.STR, 0],
        ["data", UserInfo, 0]
    ]
})
export class LoginRsp {
    /**状态码*/
    public code: number;
    /**状态信息*/
    public msg: string;
    /**用户信息*/
    public data: UserInfo;
}

/**房间*/
@MsgArray.Meta({
    id: 4,
    name: "Room",
    fields: [
        ["rid", MType.I64, 1],
        ["desc", MType.STR, 1],
        ["playing", MType.BOOL, 1],
        ["create", MType.DATE, 1],
        ["players", ["Set", UserInfo], 1],
        ["map1", ["Obj", MType.I32, UserInfo], 0],
        ["map2", ["Map", MType.I32, UserInfo], 0],
        ["log", Int8Array, 0]
    ]
})
export class Room {
    public rid: bigint;
    public desc: string;
    public playing: boolean;
    public create: Date;
    public players: Set<UserInfo>;
    public map1: {[index:number]: UserInfo};
    public map2: Map<number, UserInfo>;
    public log: Int8Array;
}
